### PR TITLE
refactor memory management

### PR DESCRIPTION
**Pull request by alexanderguzhva**
Replaced some `new T[]` with `std::make_unique<T[]>` to make the code safer.
/kind improvement

### CODE REVIEW

File: include/knowhere/dataset.h [hunk 1]
```diff
@@ -71,18 +72,46 @@ class DataSet : public std::enable_shared_from_this<const DataSet> {
...
+    void
+    SetIds(std::unique_ptr<long int[]>&& ids) {
+        static_assert(sizeof(long int) == sizeof(int64_t));
+
+        std::unique_lock lock(mutex_);
+        this->data_[meta::IDS] = Var(std::in_place_index<2>, reinterpret_cast<int64_t*>(ids.release()));
+    }
+
+    void
+    SetIds(std::unique_ptr<long long int[]>&& ids) {
```
Comment (liliu-z) [inline include/knowhere/dataset.h:108]: We we don't simply accept a `std::unique_ptr<int64_t[]>&& ids` ?

Comment (alexanderguzhva) [reply include/knowhere/dataset.h:108]: `int64_t` is different on linux and on macos, this was the root of the problem why I implemented it this way

File: include/knowhere/dataset.h [hunk 2]
```diff
@@ -71,18 +72,46 @@ class DataSet : public std::enable_shared_from_this<const DataSet> {
         this->data_[meta::DISTANCE] = Var(std::in_place_index<0>, dis);
     }
```
Comment (liliu-z) [inline include/knowhere/dataset.h:73]: We can remove these old raw pointer based API

**Review by liliu-z — APPROVED** [review]
/lgtm

**Pull request merged**